File: Greedy/3397. Maximum Number of Distinct Elements After Operations.py
class Solution:
    def maxDistinctElements(self, nums: List[int], k: int) -> int:
        nums.sort()
        n = len(nums)
        
        cur_min = nums[0] - k - 1
        for i in range(n):
            prev_range, next_range = nums[i] - k, nums[i] + k
            if cur_min < prev_range:
                cur_min = prev_range
            elif prev_range <= cur_min and cur_min < next_range:
                cur_min += 1
            nums[i] = cur_min

        return len(set(nums))

File: Sorting/3397. Maximum Number of Distinct Elements After Operations.cpp
class Solution {
public:
    int maxDistinctElements(vector<int>& nums, int k) 
    {
        int ans = 0, n = nums.size();
        sort(begin(nums), end(nums));

        int prevNumber = nums[0] - k;
        for(int i = 0; i < n; i++)
        {
            if(i and nums[i] != nums[i - 1] and !(nums[i] - k <= prevNumber)) 
                prevNumber = nums[i] - k;

            int curNumber = min(nums[i] + k, max(prevNumber, nums[i] - k));
            ans += curNumber == prevNumber;
            if(prevNumber <= nums[i] + k) prevNumber += 1;
        }
        return ans;
    }
};